packages/wordcount/src/index.js

```diff
@@ -50,56 +62,62 @@ function loadSettings( type, userSettings ) {
 }
 
 /**
- * Match the regex for the type 'words'
+ * Count the words in text
  *
  * @param {string} text     The text being processed
- * @param {string} regex    The regular expression pattern being matched
- * @param {Object} settings Settings object containing regular expressions for each strip function
+ * @param {RegExp|undefined} regex    The regular expression pattern being matched
```

Comment:
Currently, it is documented to use square-bracket notation for undefined values.

```suggestion
 * @param {RegExp} [regex]    The regular expression pattern being matched
```

>If a parameter is optional, use the square-bracket notation.

https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#nullable-undefined-and-void-types

This was left inherited / unchanged from upstream guidelines (the examples): https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/#functions

I've personally observed a few cases where TypeScript isn't interpreting the possibly-undefined type in ways I expect, where occasionally the ClosureCompiler format can work more effectively:

```suggestion
 * @param {RegExp=} regex    The regular expression pattern being matched
```

Reply:
Thanks, @aduth, and I apologize for missing this comment the other day.

From what I have experienced, there is a subtle difference between a possible `undefined` value and an optional argument. In this case, I don't think we’d like to allow _omitting_ the `regex` argument, though it would be acceptable that it's supplied and its value is `undefined`.

One side effect I have observed when I tried out either `[regex]` or `{RegExp=}` was that VSCode didn't include the `regex` when automatically filling in function arguments:

<img width="854" alt="Screenshot 2020-05-06 at 15 03 24 " src="https://user-images.githubusercontent.com/27954/81174651-e3f64000-8faa-11ea-870f-e34804e69d91.png">

This specific case is especially weird, because the third argument is **not** optional – that makes the semantic difference between `Type|undefined` and `Type=` even more pronounced.

Based on all the above, let's keep `RegExp|undefined`.

Reply:
Allowing optional parameters to precede required parameters is a bug that was recently fixed in TypeScript (unreleased): https://github.com/microsoft/TypeScript/issues/37703

This affected the `url` types: https://github.com/WordPress/gutenberg/pull/21926

---

I haven't looked closely, but do we want to allow `undefined` as the regex?